database/rrdlabels.c

```diff
@@ -3,201 +3,1138 @@
+        return 1;
+    }
+
+    return 0;
+}
+
+void rrdlabels_to_buffer(DICTIONARY *labels, BUFFER *wb, const char *before_each, const char *equal, const char *quote, const char *between_them, bool (*filter_callback)(const char *name, const char *value, RRDLABEL_SRC ls, void *data), void *filter_data, void (*name_sanitizer)(char *dst, const char *src, size_t dst_size), void (*value_sanitizer)(char *dst, const char *src, size_t dst_size)) {
+    struct labels_to_buffer tmp = {
+        .wb = wb,
+        .filter_callback = filter_callback,
+        .filter_data = filter_data,
+        .name_sanitizer = name_sanitizer,
+        .value_sanitizer = value_sanitizer,
+        .before_each = before_each,
+        .equal = equal,
+        .quote = quote,
+        .between_them = between_them,
+        .count = 0
+    };
+    dictionary_sorted_walkthrough_read(labels, label_to_buffer_callback, (void *)&tmp);
+}
+
+static int chart_label_store_to_sql_callback(const char *name, const char *value, RRDLABEL_SRC ls, void *data) {
+    RRDSET *st = (RRDSET *)data;
+    sql_store_chart_label(st->chart_uuid, (int)ls, (char *)name, (char *)value);
+    return 1;
+}
+
+void rrdset_update_rrdlabels(RRDSET *st, DICTIONARY *new_rrdlabels) {
+    if(!st->state->chart_labels)
+        st->state->chart_labels = rrdlabels_create();
+
+    if (new_rrdlabels)
+        rrdlabels_migrate_to_these(st->state->chart_labels, new_rrdlabels);
+
+    // TODO - we should also cleanup sqlite from old new_rrdlabels that have been removed
+    rrdlabels_walkthrough_read(st->state->chart_labels, chart_label_store_to_sql_callback, st);
+}
+
+// ----------------------------------------------------------------------------
+// rrdlabels unit test
+
+struct rrdlabels_unittest_add_a_pair {
+    const char *pair;
+    const char *expected_name;
+    const char *expected_value;
+    const char *name;
+    const char *value;
+    RRDLABEL_SRC ls;
+    int errors;
+};
+
+int rrdlabels_unittest_add_a_pair_callback(const char *name, const char *value, RRDLABEL_SRC ls, void *data) {
+    struct rrdlabels_unittest_add_a_pair *t = (struct rrdlabels_unittest_add_a_pair *)data;
+
+    t->name = name;
+    t->value = value;
+    t->ls = ls;
+
+    if(strcmp(name, t->expected_name) != 0) {
+        fprintf(stderr, "name is wrong, found \"%s\", expected \"%s\"", name, t->expected_name);
+        t->errors++;
+    }
+
+    if(value == NULL && t->expected_value == NULL) {
+        ;
+    }
+    else if(value == NULL || t->expected_value == NULL) {
+        fprintf(stderr, "value is wrong, found \"%s\", expected \"%s\"", value?value:"(null)", t->expected_value?t->expected_value:"(null)");
+        t->errors++;
+    }
+    else if(strcmp(value, t->expected_value) != 0) {
+        fprintf(stderr, "values don't match, found \"%s\", expected \"%s\"", value?value:"(null)", t->expected_value?t->expected_value:"(null)");
```

Comment:
I also received this warning:

```c
*** CID 353214:  Control flow issues  (DEADCODE)
/database/rrdlabels.c: 965 in rrdlabels_unittest_add_a_pair_callback()
959         }
960         else if(value == NULL || t->expected_value == NULL) {
961             fprintf(stderr, "value is wrong, found \"%s\", expected \"%s\"", value?value:"(null)", t->expected_value?t->expected_value:"(null)");
962             t->errors++;
963         }
964         else if(strcmp(value, t->expected_value) != 0) {
>>>     CID 353214:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach the expression ""(null)"" inside this statement: "fprintf(stderr, "values don...".
965             fprintf(stderr, "values don't match, found \"%s\", expected \"%s\"", value?value:"(null)", t->expected_value?t->expected_value:"(null)");
966             t->errors++;
967         }
968     
969         return 1;
970     }

** CID 353213:  Control flow issues  (DEADCODE)
/database/rrdlabels.c: 829 in rrdlabels_log_label_to_buffer_callback()
```

Reply:
This is a unit test. Yes, if there is no bug, this part will never run.
But the whole point is to catch the error if it happens.